File: cpbook/3_problem_solving_paradigms/11553_grid_game.py
# UVa 11553 - Grid Game
# https://onlinejudge.org/external/115/11553.pdf

from itertools import permutations
from sys import maxsize
import copy

memo_perm = {}

def game(n, grid, order):
	alice_candies = 0
	for i in order:
		bob_column = 0
		for j in range(1, n):
			if grid[i][j] < grid[i][bob_column]:
				bob_column = j
		alice_candies += grid[i][bob_column]
		for k in range(n):
			grid[k][bob_column] = 1001
	return alice_candies
		

def test_case():
	global memo_perm
	n = int(input())
	grid = []
	for _ in range(n):
		row = [int(x) for x in input().split()]
		grid.append(row)
	if n in memo_perm:
		perm = memo_perm[n]
	else:
		memo_perm[n] = perm = list(permutations(range(n)))
	best_result_neg = 0
	best_result_pos = 0
	# print("Case")
	for p in perm:
		g = copy.deepcopy(grid)
		result = game(n, g, p)
		# print(f"{p}: {result}")
		# best_result = max(best_result, result)
		if result < 0:
			best_result_neg = min(best_result_neg, result)
		else:
			best_result_pos = max(best_result_pos, result)
	print(best_result_neg if best_result_neg < 0 else best_result_pos)


if __name__ == "__main__":
	t = int(input())
	for _ in range(t):
		test_case()

File: cpbook/3_problem_solving_paradigms/11553_grid_game.cpp
// UVa 11553 - Grid Game
// https://onlinejudge.org/external/115/11553.pdf

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

void load_data(vector<vector<int>>& grid, int n)
{
	vector<int> row;
	int x;

	for (int i = 0; i < n; i++)	{
		row.clear();
		for (int j = 0; j < n; j++)	{
			cin >> x;
			row.push_back(x);
		}
		grid.push_back(row);
	}
}

int game(vector<vector<int>>& grid, int n, vector<int>& col)
{
	int w = 0;

	for (int i = 0; i < n; i++)	{
		w += grid[i][col[i]];	
	}

	return w;
}

void test_case(int n)
{
	vector<vector<int>> grid;
	vector<int> perm;
	int best = INT_MAX;
	int result;

	load_data(grid, n);
	for (int i = 0; i < n; i++)	{
		perm.push_back(i);
	}
	do	{
		result = game(grid, n, perm);
		best = min(best, result);
	}	while (next_permutation(perm.begin(), perm.end()));
	cout << best << endl;
}


int main()
{
	int tc, n;

	ios::sync_with_stdio(0);
	cin.tie(0);
	cin >> tc;
	while (tc--)	{
		cin >> n;
		test_case(n);
	}
	return 0;
}
